pkg/pillar/cmd/downloader/context.go

```diff
@@ -0,0 +1,181 @@
+	ctx.subGlobalDownloadConfig = subGlobalDownloadConfig
+	subGlobalDownloadConfig.Activate()
+
+	// Look for DatastoreConfig. We should process this
+	// before any download config ( App/baseos/cert). Without DataStore Config,
+	// Image Downloads will run into errors.
+	subDatastoreConfig, err := pubsub.Subscribe("zedagent",
+		types.DatastoreConfig{}, false, &ctx)
+	if err != nil {
+		return err
+	}
+	subDatastoreConfig.ModifyHandler = handleDatastoreConfigModify
+	subDatastoreConfig.CreateHandler = handleDatastoreConfigModify
+	subDatastoreConfig.DeleteHandler = handleDatastoreConfigDelete
+	ctx.subDatastoreConfig = subDatastoreConfig
+	subDatastoreConfig.Activate()
+
+	pubGlobalDownloadStatus, err := pubsub.Publish(agentName,
+		types.GlobalDownloadStatus{})
+	if err != nil {
+		return err
+	}
+	ctx.pubGlobalDownloadStatus = pubGlobalDownloadStatus
+
+	// Set up our publications before the subscriptions so ctx is set
+	pubAppImgStatus, err := pubsub.PublishScope(agentName, types.AppImgObj,
+		types.DownloaderStatus{})
+	if err != nil {
+		return err
+	}
+	ctx.pubAppImgStatus = pubAppImgStatus
+	pubAppImgStatus.ClearRestarted()
+
+	pubBaseOsStatus, err := pubsub.PublishScope(agentName, types.BaseOsObj,
+		types.DownloaderStatus{})
+	if err != nil {
+		return err
+	}
+	ctx.pubBaseOsStatus = pubBaseOsStatus
+	pubBaseOsStatus.ClearRestarted()
+
+	pubCertObjStatus, err := pubsub.PublishScope(agentName, types.CertObj,
+		types.DownloaderStatus{})
+	if err != nil {
+		return err
+	}
+	ctx.pubCertObjStatus = pubCertObjStatus
+	pubCertObjStatus.ClearRestarted()
+
+	subAppImgConfig, err := pubsub.SubscribeScope("zedmanager",
+		types.AppImgObj, types.DownloaderConfig{}, false, &ctx)
+	if err != nil {
+		return err
+	}
+	subAppImgConfig.ModifyHandler = handleAppImgModify
+	subAppImgConfig.CreateHandler = handleAppImgCreate
+	subAppImgConfig.DeleteHandler = handleAppImgDelete
+	ctx.subAppImgConfig = subAppImgConfig
+	subAppImgConfig.Activate()
+
+	subBaseOsConfig, err := pubsub.SubscribeScope("baseosmgr",
+		types.BaseOsObj, types.DownloaderConfig{}, false, &ctx)
+	if err != nil {
+		return err
+	}
+	subBaseOsConfig.ModifyHandler = handleBaseOsModify
+	subBaseOsConfig.CreateHandler = handleBaseOsCreate
+	subBaseOsConfig.DeleteHandler = handleBaseOsDelete
+	ctx.subBaseOsConfig = subBaseOsConfig
+	subBaseOsConfig.Activate()
+
+	subCertObjConfig, err := pubsub.SubscribeScope("baseosmgr",
+		types.CertObj, types.DownloaderConfig{}, false, &ctx)
+	if err != nil {
+		return err
+	}
+	subCertObjConfig.ModifyHandler = handleCertObjModify
+	subCertObjConfig.CreateHandler = handleCertObjCreate
+	subCertObjConfig.DeleteHandler = handleCertObjDelete
+	ctx.subCertObjConfig = subCertObjConfig
+	subCertObjConfig.Activate()
+
+	pubAppImgStatus.SignalRestarted()
+	pubBaseOsStatus.SignalRestarted()
+	pubCertObjStatus.SignalRestarted()
+
+	return nil
+}
+
+func (ctx *downloaderContext) subscription(objType string) *pubsub.Subscription {
```

Comment:
I think I'd prefer longer names for this in case we apply this approach to other agents.
Calling the function downloaderSubscription (or DownloadConfigSubscription) makes it more clear that it isn't a way to access any of the subscriptions the agent is using.

Reply:
The context is called a `downloaderContext`, so by definition any `subscription()` call on it is going to be a downloader subscription (which is why go linters often complain about that).

This part has me a bit confused though:

> more clear that it isn't a way to access any of the subscriptions the agent is using

Isn't that what it is doing? If I call `ctx.subscription(types.AppImgObj)`, it returns `ctx.subAppImgConfig`, which is a `*pubsub.Subscription`.

Reply:
downloader is an agent.
DownloadConfig is a set of input to that agent.
Based on your current naming it seems like it is about subscriptions for the agent.
Note that the agent subscribes to other things like GlobalConfig, DeviceNetworkStatus, etc.

Reply:
I am still somewhat confused. Downloader is an agent, ok. `downloaderContext` (both before and after this PR, as the struct itself is unchanged) is a holder for all of the `Subscription`s and `Publication`s for the agent. Does this function - formerly named `downloaderPublication()`, proposed `downloaderContext.publication()` - not retrieve a publication on the agent? `downloaderPublication()` beforehand just retrieved a named publication from the context, and `downloaderContext.publication()` does the same?

Reply:
@eriknordmark are we good to go, based on our slack discussion?

I will rebase right away.

Reply:
Yes, as I said on slack I hadn't realized that the name downloaderSubscription() was the original name.